src/compiler/checker.ts

```diff
@@ -8405,16 +8405,19 @@ namespace ts {
             if (forEach(node.parameters, p => !getEffectiveTypeAnnotationNode(p))) {
                 return true;
             }
-            // For arrow functions we now know we're not context sensitive.
-            if (node.kind === SyntaxKind.ArrowFunction) {
-                return false;
+            if (node.kind !== SyntaxKind.ArrowFunction) {
+                // If the first parameter is not an explicit 'this' parameter, then the function has
+                // an implicit 'this' parameter which is subject to contextual typing. Otherwise we
+                // know that all parameters (including 'this') have type annotations and nothing is

```

Comment:
"nothing is subject to contextual typing" should change to read "only the return type could still be subject to contextual typing".